src/util/ConcurrentCache.h

```diff
@@ -0,0 +1,296 @@
+      : _cacheAndInProgressMap{std::forward<CacheArgs>(cacheArgs)...} {}
+
+  struct ResultAndCacheStatus {
+    shared_ptr<const Value> _resultPointer;
+    bool _wasCached;
+  };
+
+ public:
+  /**
+   * @brief Obtain the result of an expensive computation. Do not recompute the
+   * result if it is cached or currently being computed by another thread.
+   * @tparam ComputeFunction A callable whose operator() takes no argument and
+   * produces the computation result.
+   * @param key  A key that can uniquely identify a computation. For equal keys,
+   * the associated computeFunctions must yield the same results.
+   * @param computeFunction The actual computation. If the result has to be
+   * computed, computeFunction() is called.
+   * @return A shared_ptr to the computation result.
+   *
+   */
+  template <class ComputeFunction>
+  ResultAndCacheStatus computeOnce(const Key& key,
+                                   ComputeFunction computeFunction) {
+    return computeOnceImpl(false, key, std::move(computeFunction));
+  }
+
+  /// Similar to computeOnce, with the following addition: After the call
+  /// completes, the result will be pinned in the underlying cache.
+  template <class ComputeFunction>
+  ResultAndCacheStatus computeOncePinned(const Key& key,
+                                         ComputeFunction computeFunction) {
+    return computeOnceImpl(true, key, std::move(computeFunction));
+  }
+
+  /// Clear the cache (but not the pinned elements)
+  void clear() { _cacheAndInProgressMap.wlock()->_cache.clearUnpinnedOnly(); }
+
+  /// Clear the cache, including the pinned elements.
+  void clearAll() { _cacheAndInProgressMap.wlock()->_cache.clearAll(); }
+
+  /// The number of non-pinned elements in the cache
+  auto numCachedElements() const {
+    return _cacheAndInProgressMap.wlock()->_cache.numCachedElements();
+  }
+
+  /// The number of pinned elements in the underlying cache
+  auto numPinnedElements() const {
+    return _cacheAndInProgressMap.wlock()->_cache.numPinnedElements();
+  }
+
+  /// Total size of the non-pinned elements in the cache (the unit depends on
+  /// the cache's configuration)
+  auto cachedSize() const {
+    return _cacheAndInProgressMap.wlock()->_cache.cachedSize();
+  }
+
+  /// Total size of the non-pinned elements in the cache (the unit depends on
+  /// the cache's configuration)
+  auto pinnedSize() const {
+    return _cacheAndInProgressMap.wlock()->_cache.pinnedSize();
+  }
+
+  /// only for testing: get access to the implementation
+  auto& getStorage() { return _cacheAndInProgressMap; }
+
+  // is key in cache (not in progress), used for testing
+  bool cacheContains(const Key& k) const {
+    return _cacheAndInProgressMap.wlock()->_cache.contains(k);
+  }
+
+  // Get element from cache by its key.
+  // The behavior if the key is not present in the cache depends on the
+  // cache type's operator[]. The ad_utility caches will return a nullptr
```

Comment:
// If the key is not present in the cache, the behavior depends on the behavior of operator[] of _cache. The ad_utility Cache return a nullptr in that case. The return type is shared_ptr<const Value>.

Omit the newline after the first sentence of the comment.